sfeed_frames.1
.Dd August 16, 2018
.Dt SFEED_FRAMES 1
.Os
.Sh NAME
.Nm sfeed_frames
.Nd format feed data to HTML with frames
.Sh SYNOPSIS
.Nm
.Op Ar file...
.Sh DESCRIPTION
.Nm
formats feed data (TSV) from
.Xr sfeed 1
to HTML.
It reads TSV data from stdin or
.Ar file
and writes HTML files for the frameset to the current directory.
If no
.Ar file
parameters are specified and therefore the data is read from stdin then the
menu.html file is not written.
.Sh FILES WRITTEN
.Bl -tag -width 13n
.It index.html
The main HTML file referencing to the frames items.html and menu.html.
.It items.html
The items frame contains all the item HTML links to the remote content.
.It menu.html
The menu frame which contains navigation "anchor" links to the feed names in
items.html.
.El
.Sh SEE ALSO
.Xr sfeed 1 ,
.Xr sfeed_html 1 ,
.Xr sfeed_plain 1
.Sh AUTHORS
.An Hiltjo Posthuma Aq Mt hiltjo@codemadness.org


CHANGELOG
v1.0 (unreleased)
====

* Simplified the sfeed(5) format:
  * The formatted time field is removed.
  * The feedtype field (values "atom" or "rss") is removed.
  To convert your files from the previous format to the changed format you
  can just remove the fields:
      cd ~/.sfeed/feeds
      for i in *; do cut -f 1,3-8 "$i" > /tmp/t; mv /tmp/t "$i"; done
* Simpler and much faster time parsing in sfeed(1): like timegm(3), this is
  also more portable.
* 2038-proof: support 64-bit time_t, output truncated as long long int.
* The format programs now use the localtime and a shorter time format:
  "yyyy-mm-dd HH:MM" instead of "yyyy-mm-dd HH:MM:SS timezone".
  If you run the programs on a system in a different timezone make sure to
  update the TZ environment variable or change the source-code to show the UTC
  offset.

Minor fixes:
* Check errno (ERANGE) when parsing UNIX timestamp from a sfeed(5) file.
* sfeed_frames: don't make title of content a HTML anchor if it's empty.
* sfeed_frames: make the filename of the content file more unique: use the
  UNIX timestamp in the filename. This makes sure it won't overwrite older
  content files anymore which had the same title, but a different timestamp.

Documentation:
* Add sfeed(5) man page which explains the feed file format, this information
  is also still available in sfeed(1).
* Rewrite some sentences so it is hopefully more clear.


v0.9.1
======

* Improved `make dist` rule in Makefile: don't build binaries before packaging
  the source files. Package the source files in the relative path
  sfeed-VERSION/.


v0.9
====

Features:
---------

* sfeed_update: feeds are now by default updated concurrently for an increase
  in performance.
* sfeed_update: Auto-detect XML character set encoding (sfeed_xmlenc).
* sfeed_update: don't send user-agent by default (used to send: curl+version).
* Removed the dependency on libexpat, using a custom XML parser (xml.*),
  this parser is also non-validating, it will not check the XML for errors
  or stop parsing if it contains errors (which is good).
* On OpenBSD 5.9+ the sfeed tools can now use pledge(2) as a security
  mitigation. Most tools only require "stdio" and read-only filesystem access.
  To enable it the CPPFLAGS -DUSE_PLEDGE can be used, see the file: config.mk.
* Added (optional) format tools:
  * Add sfeed_frames, a formatting program to output feeds as a HTML file with
    frames. It's optimized to look good in older browsers that don't necesarily
    support CSS or modern HTML, like links. See the man page for more details.
  * Add sfeed_mbox: format feeds to mbox format.
  * Add sfeed_tail: poll files or stdin and print unique/unseen items to
    stdout, this is inspired from RSSTail.
  * Add sfeed_web: find RSS/Atom feeds from web-pages.


Documentation:
--------------

* Changed to mandoc (semantic) for the documentation format, for older systems
  man pages can be created with mandoc -Tman.
* Major improvements to the documentation quality.


License:
--------

* Changed license from MIT to ISC (simpler wording).


Bug fixes:
----------

* sfeed_update: fix issue with merging failed feeds, the date from the last
  (possibly failed) request would be used and passed as HTTP If-Modified-Since,
  so data would be missed, see commit 77a603a904087dd9fd3350da029f279f076e4f4b
  for more info.
* Fix possible TOCTOU in sfeed_frames.
* Fix crash with repeated tags: <link href="" /> in Atom feeds.
* Path truncations are now treated as fatal errors.
* Remove use of deprecated (POSIX) functions: timegm, utime.
* Fix off-by-one in string_append when reallocating buffer.
* Parsing datetime strings containing timezone suffix to UNIX timestamps
  was wrong. UNIX timestamps are in UTC+0 time. For simplicity sake output
  parsed datetime strings in a formatted manner without converting the
  time to the user selected timezone YYYY-mm-dd HH:MM:SS (UTC[+-][HHMM])|tz .
* Also to determine if an item is new - by comparing the UNIX timestamp -
  don't take into account timezones, because setenv(TZ,...) is just ugly and
  I don't use it.
* Lots of small fixes for compatibility with RSS / Atom feeds.
* Many more improvements, I'll keep a more detailed list in the future.


v0.8
====

No history


sfeed_frames.c
#include <sys/stat.h>
#include <sys/types.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "util.h"

static struct feed **feeds;
static char *line;
static size_t linesize;
static time_t comparetime;
static unsigned long totalnew;

static void
printfeed(FILE *fpitems, FILE *fpin, struct feed *f)
{
	char *fields[FieldLast];
	ssize_t linelen;
	unsigned int isnew;
	struct tm *tm;
	time_t parsedtime;

	/* menu if not unnamed */
	if (f->name[0]) {
		fputs("<h2 id=\"", fpitems);
		xmlencode(f->name, fpitems);
		fputs("\"><a href=\"#", fpitems);
		xmlencode(f->name, fpitems);
		fputs("\">", fpitems);
		xmlencode(f->name, fpitems);
		fputs("</a></h2>\n", fpitems);
	}

	while ((linelen = getline(&line, &linesize, fpin)) > 0) {
		if (line[linelen - 1] == '\n')
			line[--linelen] = '\0';
		if (!parseline(line, fields))
			break;

		parsedtime = 0;
		if (strtotime(fields[FieldUnixTimestamp], &parsedtime))
			continue;
		if (!(tm = localtime(&parsedtime)))
			err(1, "localtime");

		isnew = (parsedtime >= comparetime) ? 1 : 0;
		totalnew += isnew;
		f->totalnew += isnew;
		f->total++;

		fprintf(fpitems, "%04d-%02d-%02d&nbsp;%02d:%02d ",
		        tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
		        tm->tm_hour, tm->tm_min);

		if (isnew)
			fputs("<b><u>", fpitems);
		if (fields[FieldLink][0]) {
			fputs("<a href=\"", fpitems);
			xmlencode(fields[FieldLink], fpitems);
			fputs("\">", fpitems);
			xmlencode(fields[FieldTitle], fpitems);
			fputs("</a>", fpitems);
		} else {
			xmlencode(fields[FieldTitle], fpitems);
		}
		if (isnew)
			fputs("</u></b>", fpitems);
		fputs("\n", fpitems);
	}
}

int
main(int argc, char *argv[])
{
	FILE *fpindex, *fpitems, *fpmenu = NULL, *fp;
	char *name;
	int i, showsidebar = (argc > 1);
	struct feed *f;

	if (pledge("stdio rpath wpath cpath", NULL) == -1)
		err(1, "pledge");

	if (!(feeds = calloc(argc, sizeof(struct feed *))))
		err(1, "calloc");

	if ((comparetime = time(NULL)) == -1)
		err(1, "time");
	/* 1 day is old news */
	comparetime -= 86400;

	/* write main index page */
	if (!(fpindex = fopen("index.html", "wb")))
		err(1, "fopen: index.html");
	if (!(fpitems = fopen("items.html", "wb")))
		err(1, "fopen: items.html");
	if (showsidebar && !(fpmenu = fopen("menu.html", "wb")))
		err(1, "fopen: menu.html");

	if (pledge(argc == 1 ? "stdio" : "stdio rpath", NULL) == -1)
		err(1, "pledge");

	fputs("<html><head><link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\" />"
	      "<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\" /></head>"
	      "<body class=\"frame\"><div id=\"items\"><pre>", fpitems);

	if (argc == 1) {
		if (!(feeds[0] = calloc(1, sizeof(struct feed))))
			err(1, "calloc");
		feeds[0]->name = "";
		printfeed(fpitems, stdin, feeds[0]);
	} else {
		for (i = 1; i < argc; i++) {
			if (!(feeds[i - 1] = calloc(1, sizeof(struct feed))))
				err(1, "calloc");
			name = ((name = strrchr(argv[i], '/'))) ? name + 1 : argv[i];
			feeds[i - 1]->name = name;

			if (!(fp = fopen(argv[i], "r")))
				err(1, "fopen: %s", argv[i]);
			printfeed(fpitems, fp, feeds[i - 1]);
			if (ferror(fp))
				err(1, "ferror: %s", argv[i]);
			fclose(fp);
		}
	}
	fputs("</pre>\n</div></body>\n</html>\n", fpitems); /* div items */

	if (showsidebar) {
		fputs("<html><head>"
		      "<link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\" />\n"
		      "<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\" />\n"
		      "</head><body class=\"frame\"><div id=\"sidebar\">", fpmenu);

		for (i = 1; i < argc; i++) {
			f = feeds[i - 1];
			if (f->totalnew)
				fputs("<a class=\"n\" href=\"items.html#", fpmenu);
			else
				fputs("<a href=\"items.html#", fpmenu);
			xmlencode(f->name, fpmenu);
			fputs("\" target=\"items\">", fpmenu);
			if (f->totalnew > 0)
				fputs("<b><u>", fpmenu);
			xmlencode(f->name, fpmenu);
			fprintf(fpmenu, " (%lu)", f->totalnew);
			if (f->totalnew > 0)
				fputs("</u></b>", fpmenu);
			fputs("</a><br/>\n", fpmenu);
		}
		fputs("</div></body></html>\n", fpmenu);
	}
	fputs("<!DOCTYPE html><html><head>\n\t<title>Newsfeed (", fpindex);
	fprintf(fpindex, "%lu", totalnew);
	fputs(")</title>\n\t<link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\" />\n"
	      "<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\" />\n"
	      "</head>\n", fpindex);
	if (showsidebar) {
		fputs("<frameset framespacing=\"0\" cols=\"250,*\" frameborder=\"1\">\n"
		      "\t<frame name=\"menu\" src=\"menu.html\" target=\"menu\">\n", fpindex);
	} else {
		fputs("<frameset framespacing=\"0\" cols=\"*\" frameborder=\"1\">\n", fpindex);
	}
	fputs(
	      "\t<frame name=\"items\" src=\"items.html\" target=\"items\">\n"
	      "</frameset>\n"
	      "</html>\n", fpindex);

	fclose(fpindex);
	fclose(fpitems);
	if (fpmenu)
		fclose(fpmenu);

	return 0;
}
